test/common/buffer/owned_impl_test.cc

```diff
@@ -347,6 +347,164 @@ TEST_F(OwnedImplTest, Read) {
+  EXPECT_EQ("123",
+            absl::string_view(reinterpret_cast<const char*>(slice_data.data()), slice_data.size()));
+  EXPECT_EQ(buffer.toString(), "Abcde");
+}
+
+TEST_F(OwnedImplTest, ExtractAfterSentinelDiscard) {
+  // Create a buffer with a sentinel and one owned slice.
+  Buffer::OwnedImpl buffer;
+  bool sentinel_discarded = false;
+  const Buffer::OwnedBufferFragmentImpl::Releasor sentinel_releasor{
+      [&](const Buffer::OwnedBufferFragmentImpl* sentinel) {
+        sentinel_discarded = true;
+        delete sentinel;
+      }};
+  auto sentinel =
+      Buffer::OwnedBufferFragmentImpl::create(absl::string_view("", 0), sentinel_releasor);
+  buffer.addBufferFragment(*sentinel.release());
+
+  buffer.appendSliceForTest("abcde");
+  const uint64_t expected_length = 5;
+  EXPECT_EQ(buffer.toString(), "abcde");
+  RawSliceVector slices = buffer.getRawSlices(); // only returns slices with data
+  EXPECT_EQ(1, slices.size());
+
+  // Extract owned slice after discarding sentinel.
+  EXPECT_FALSE(sentinel_discarded);
+  auto slice = buffer.extractMutableFrontSlice();
+  ASSERT_TRUE(slice);
+  EXPECT_TRUE(sentinel_discarded);
+  auto slice_data = slice->getMutableData();
+  ASSERT_NE(slice_data.data(), nullptr);
+  EXPECT_EQ(slice_data.size(), expected_length);
+  EXPECT_EQ("abcde",
+            absl::string_view(reinterpret_cast<const char*>(slice_data.data()), slice_data.size()));
+  EXPECT_EQ(0, buffer.length());
+}
+
+TEST_F(OwnedImplTest, DrainThenExtractOwnedSlice) {
+  // Create a buffer with two owned slices.
+  Buffer::OwnedImpl buffer;
+  buffer.appendSliceForTest("abcde");
+  const uint64_t expected_length0 = 5;
+  buffer.appendSliceForTest("123");
+  EXPECT_EQ(buffer.toString(), "abcde123");
+  RawSliceVector slices = buffer.getRawSlices();
+  EXPECT_EQ(2, slices.size());
+
+  // Partially drain the first slice.
+  const uint64_t partial_drain_size = 2;
+  buffer.drain(partial_drain_size);
+  EXPECT_EQ(buffer.toString(), static_cast<const char*>("abcde123") + partial_drain_size);
+
+  // Extracted partially drained first slice, leaving the second slice.
+  auto slice = buffer.extractMutableFrontSlice();
+  ASSERT_TRUE(slice);
+  auto slice_data = slice->getMutableData();
+  ASSERT_NE(slice_data.data(), nullptr);
+  EXPECT_EQ(slice_data.size(), expected_length0 - partial_drain_size);
+  EXPECT_EQ(static_cast<const char*>("abcde") + partial_drain_size,
+            absl::string_view(reinterpret_cast<const char*>(slice_data.data()), slice_data.size()));
+  EXPECT_EQ(buffer.toString(), "123");
+}
+
+TEST_F(OwnedImplTest, ExtractUnownedSlice) {
+  // Create a buffer with an unowned slice.
+  std::string input{"unowned test slice"};
+  const size_t expected_length0 = input.size();
+  auto frag = OwnedBufferFragmentImpl::create(
+      {input.c_str(), expected_length0},
+      [this](const OwnedBufferFragmentImpl*) { release_callback_called_ = true; });
+  Buffer::OwnedImpl buffer;
+  buffer.addBufferFragment(*frag);
```

Comment:
Could we add a drain tracker here to verify that the extract on the immutable slice below also executes the drain tracker at the right time?